File: Example/src/main/java/com/example/game/mods/Prop.java
package com.example.game.mods;

import org.elkoserver.feature.basicexamples.cartesian.CartesianPosition;
import org.elkoserver.foundation.json.JSONMethod;
import org.elkoserver.foundation.json.MessageHandlerException;
import org.elkoserver.json.EncodeControl;
import org.elkoserver.json.JSONLiteral;
import org.elkoserver.server.context.*;

public class Prop extends Mod implements ItemMod {
    private String myKind;

    private static final int GRAB_DISTANCE = 5;

    @JSONMethod({ "kind" })
    public Prop(String kind) {
        myKind = kind;
    }

    public JSONLiteral encode(EncodeControl control) {
        JSONLiteral result = new JSONLiteral("prop", control);
        result.addParameter("kind", myKind);
        result.finish();
        return result;
    }

    @JSONMethod
    public void grab(User from) throws MessageHandlerException {
        ensureInContext(from);
        Item item = (Item) object();
        if (!item.isPortable()) {
            throw new MessageHandlerException(
                "attempt to grab non-portable item " + item);
        }

        CartesianPosition userPos = from.getMod(CartesianPosition.class);
        if (userPos == null) {
            throw new MessageHandlerException("user " + from +
                    " attempted grab " +
                    this + " but Cartesian position mod not present on user");
        }
        CartesianPosition itemPos = item.getMod(CartesianPosition.class);
        if (itemPos == null) {
            throw new MessageHandlerException("user " + from +
                    " attempted grab " +
                    this + " but Cartesian position mod not present on item");
        }
        int dx = itemPos.x() - userPos.x();
        int dy = itemPos.y() - userPos.y();
        if (dx*dx + dy*dy > GRAB_DISTANCE*GRAB_DISTANCE) {
            throw new MessageHandlerException(
                "attempt to grab too far away item " + item);
        }

        item.setContainer(from);
        itemPos.detach();
        context().sendToNeighbors(from, Msg.msgDelete(item));
        from.send(Movement.msgMove(item, 0, 0, from));
    }

    @JSONMethod
    public void drop(User from) throws MessageHandlerException {
        ensureHolding(from);
        Item item = (Item) object();
        if (!item.isPortable()) {
            throw new MessageHandlerException(
                "attempt to drop non-portable item " + item);
        }
        CartesianPosition userPos = from.getMod(CartesianPosition.class);
        if (userPos == null) {
            throw new MessageHandlerException("user " + from +
                    " attempted drop " +
                    this + " but Cartesian position mod not present on user");
        }
        CartesianPosition itemPos = item.getMod(CartesianPosition.class);
        if (itemPos == null) {
            itemPos = new CartesianPosition(userPos.x(), userPos.y());
            itemPos.attachTo(item);
        } else {
            itemPos.set(userPos.x(), userPos.y());
        }
        item.setContainer(context());
        item.sendObjectDescription(context().neighbors(from), context());
        from.send(Movement.msgMove(item, userPos.x(), userPos.y(), context()));
    }
}

File: Example/src/main/kotlin/com/example/game/mods/Prop.kt
package com.example.game.mods

import org.elkoserver.feature.basicexamples.cartesian.CartesianPosition
import org.elkoserver.foundation.json.JsonMethod
import org.elkoserver.foundation.json.MessageHandlerException
import org.elkoserver.json.EncodeControl
import org.elkoserver.json.JsonLiteral
import org.elkoserver.json.JsonLiteralFactory
import org.elkoserver.server.context.Item
import org.elkoserver.server.context.ItemMod
import org.elkoserver.server.context.Mod
import org.elkoserver.server.context.User
import org.elkoserver.server.context.msgDelete

class Prop @JsonMethod("kind") constructor(private val myKind: String) : Mod(), ItemMod {
    override fun encode(control: EncodeControl): JsonLiteral =
            JsonLiteralFactory.type("prop", control).apply {
                addParameter("kind", myKind)
                finish()
            }

    @JsonMethod
    fun grab(from: User) {
        ensureInContext(from)
        val item = `object`() as Item
        assertIsPortable()
        val userPos = getUserPosition(from)
        val itemPos = getItemPosition(item, from)
        val dx = itemPos.x - userPos.x
        val dy = itemPos.y - userPos.y
        assertWithinGrabDistance(dx, dy, item)
        item.setContainer(from)
        itemPos.detach()
        context().sendToNeighbors(from, msgDelete(item))
        from.send(msgMove(item, 0, 0, from))
    }

    @JsonMethod
    fun drop(from: User) {
        ensureHolding(from)
        val item = `object`() as Item
        assertIsPortable()
        val userPos = getUserPosition(from)
        var itemPos = item.getMod(CartesianPosition::class.java)
        if (itemPos == null) {
            itemPos = CartesianPosition(userPos.x, userPos.y)
            itemPos.attachTo(item)
        } else {
            itemPos[userPos.x] = userPos.y
        }
        item.setContainer(context())
        item.sendObjectDescription(context().neighbors(from), context())
        from.send(msgMove(item, userPos.x, userPos.y, context()))
    }

    private fun assertIsPortable() {
        if (!(`object`() as Item).isPortable) {
            throw MessageHandlerException("attempt to grab/drop non-portable item ${`object`()}")
        }
    }

    private fun getUserPosition(from: User) =
            from.getMod(CartesianPosition::class.java)
                    ?: throw MessageHandlerException("user $from attempted grab/drop $this but Cartesian position mod not present on user")

    private fun getItemPosition(item: Item, from: User) =
            item.getMod(CartesianPosition::class.java)
                    ?: throw MessageHandlerException("user $from attempted grab $this but Cartesian position mod not present on item")

    private fun assertWithinGrabDistance(dx: Int, dy: Int, item: Item) {
        if (dx * dx + dy * dy > GRAB_DISTANCE * GRAB_DISTANCE) {
            throw MessageHandlerException("attempt to grab too far away item $item")
        }
    }

    companion object {
        private const val GRAB_DISTANCE = 5
    }
}
